### ISSUE TITLE
How do I make a multiband compressor?

### ISSUE THREAD

**Issue by elanhickler** [RobinSchmidt/RS-MET#121]
Hey Robin, do you have the components in your library to create a multiband compressor? What components should i use? If you don't have the components I need to commission you to make them.

**Comment by RobinSchmidt** [comment]
how many bands do you need? if it's no more than 4, then i'd say, you could use my 4-way linkwitz/riley crossover and apply a compressor to each band (there's a compressor class in rosic). if you need more than 4 bands, we may have to develop a multiband-crossover for that. ...and/or if you need perfect reconstruction, we may have to create some perfect-recosntruction band-splitter first. it's easy to do with FIR filters but not so easy with IIR filters. i once have more or less successfully experimented with 12 dB/oct perfect-reconstruction splitters. i may have to dig that out and continue where i left off

**Comment by RobinSchmidt** [comment]
perfect reconstruction in this context means that if you split your input into a lowpass and a highpass part and add both parts together, you get your original signal back. that's sort of hard to do with high-order (i.e. steep) IIR filters. ehen you ad a linkwitz/riley splitted lowpass/highpass back together, you get an allpassed version of the input. that's already a resonably good compromise. if you use just any old pair of highpass/lowpass filter and add them, you get a mess (dips and peaks in the combined magnitude response)

**Comment by RobinSchmidt** [comment]
....and if you think, you could just obtain your highpass signal by subtracting the lowpass-signal from the original signal, look at the magnitude response of the resulting "highpass". in this case, *this* will be a mess

**Comment by RobinSchmidt** [comment]
so, bottom line, if up-to 4 bands and allpass reconstructions is good enough, you can use my linkwitz/riley splitter. if perfect reconstruction and/or more bands are needed, we may have to look into developing appropriate band-splitters first. once you have a perfect-reconstruction lowpass/highpass pair, it's trivial to make a multiband splitter by just using that prototype splitter pair in a binary tree, i.e. low and high bands get split further into low-low, low-high, high-low, high-high and so on

**Comment by elanhickler** [comment]
I need unlimited number of bands, but it could be around 4.

What can you get done quickly?

Would it be easy to allow more than 4 bands on the linkwitz/riley splitter? Or does the algorithm have a limitation of 4? Going to 5 or 6 might not be so bad.

**Comment by RobinSchmidt** [comment]
i think, really quick would be only using 6 dB/oct splitting filters. they provide perfect reconstruction by nature. anything else would require some R&D. extending the linkwitz/riley splitter to 8 bands (i.e. another level in the tree) is possible in principle, but it would be *very* messy because you would have to consider so many combinations of how lower-levels can be split further (or not). the number of possibilities explodes exponentially with the number of levels

**Comment by RobinSchmidt** [comment]
maybe read the user manual of my crossover:
http://www.rs-met.com/software/freebies/CrossOver_v10.04.zip
i explain the issues a bit in the last section

**Comment by RobinSchmidt** [comment]
it will be a mess having to deal with so many possibilities for the required compensation allpasses

**Comment by RobinSchmidt** [comment]
hmmm...."First-order filters are considered by many audiophiles to be ideal for crossovers."

https://en.wikipedia.org/wiki/Audio_crossover#First_order

perhaps the best solution would be to try to make my 12 dB/oct perfect-reconstruction splitter work. it would combine the advantage of perfect-reconstruction with a reasonable band separation.

**Comment by elanhickler** [comment]
... I need a good balance of time it would take to results. So what would you choose?

**Comment by RobinSchmidt** [comment]
well, as said - if you need something quick, using 1st order filters would be the fastest route to go (you would just need to compute the lowpass and obtain the highpass by subtraction from the original). build a binary tree from these for multiband splitting with an arbitrary number of bands. i would perhaps try to set the structure up in such a way that they can be later easily replaced (at runtime, as an user option) with other filters (say the 12 dB ones, or maybe FIR).

**Comment by elanhickler** [comment]
Can you get the FIR splutter done in 2 weeks?

**Comment by RobinSchmidt** [comment]
when there aren't any unforeseen obstacles, probably yes. i just wonder, how the programming interface for accessing the tree-structure would look like. i guess, that would be the most complicated part and perhaps inconvenient from a user's perspective who may rather want to think about it as a parallel filter bank.

also - do you need the cutoff/crossover frequencies to be modulatable? because that would be impractical with FIR filters. an FIR approach would only be good for fixed filters. it's basically a convolution for each lowpass/highpass pair and changing a cutoff frequencies requires recomputation of the impulse response, which is expensive

**Comment by elanhickler** [comment]
yes fixed! go for it.

**Comment by RobinSchmidt** [comment]
i have been thinking about this some more and i have now a whole bunch of ideas (maybe see the recently updated ToDo.txt) ...but in any case, i think, i'll need to drag over and adapt/templatize etc. some more rosic classes into rapt before. i want to develop the new stuff in the context of the rapt library. so , i think, i'll use the coming days to drag over all the classes that may be relevant for this task. i won't charge you for this. while doing this, i can continue thinking about what would be the best approach. i'm not really sure, if it will be done in two weeks though. it's really some R&D task

**Comment by RobinSchmidt** [comment]
i think the programming interface should have functions like:

```
setNumberOfBands(int N) 
setCrossoverFrequency(int band, double newCutoff) 
processSampleFrame(double in, double* outs)
```

i think, that would be convenient to use

**Comment by RobinSchmidt** [comment]
i guess, it will be one of the two options:

(1) FFT, spectral-splitting, several iFFTs
(2) bandpass-bank with compensation (flattening) filter

somehow i think, (2) is more elegant. it's the analog way and doesn't introduce block-processing (which i consider somehow unnatural). on the other hand, (1) could give perfect reconstruction whereas (2) would probably at best give allpass reconstruction. hmmm...

**Comment by RobinSchmidt** [comment]
the idea in (2) is that we would just use some parallel connection of bandpass filters (perhaps butterworth of butterworth-squared) and whatever wiggle there will be in the frequency-response of the recombination of the bands may possibly be compensated for by some other filter that can be applied pre-splitting or post-recombination

**Comment by elanhickler** [comment]
I don't have an opinion on that

**Comment by RobinSchmidt** [comment]
o, i think, i'll first drag over my EngineersFilter stuff (i need that anyway - it's an important building block of many dsp algorithms) and try to experiment with a filter-bank made of some engineers-filter powered bandpasses. there certainly will be some wiggle in the recombined frequency-response, but it could be interesting to figure out, which kind of filters give the least amount of it. and trying to compensate for whatever wiggle remains could lead to some interesting math.

**Comment by RobinSchmidt** [comment]
i'll probably have to find the poles and zeros of the recombined frequency response and place compensation poles/zeros into the z-plane. that going to be some nerdy fun!

**Comment by RobinSchmidt** [comment]
and i think, i'll create some new tools in the process - such as plotting pole/zero patterns of filters in GNUPlot, which will be useful for many other things

**Comment by RobinSchmidt** [comment]
just bookmarking some possibly relevant papers:
http://www.dafx17.eca.ed.ac.uk/papers/DAFx17_paper_94.pdf
https://www.researchgate.net/publication/265789445_High-Precision_Parallel_Graphic_Equalizer

...lots of boring copy/paste/edit grunt-work to do to drag over all my IIR filter stuff into rapt, but i see light at the end of the tunnel

**Comment by RobinSchmidt** [comment]
i think, i'll try the perfect-reconstruction splitter-tree first. if that works out, we'll have a perfect splitter

**Comment by elanhickler** [comment]
how about gui? You don't have to create a GUI that shows the bands and you don't have to ensure that bands don't overlap. Getting it done on time is more important.

**Comment by RobinSchmidt** [comment]
gui-wise, i'd ideally imagine something like a mixture of my EasyQ and CrossOver plugin: you may freely insert new bandsplits by mouse-clicks like you insert more eq-stages in easyq, but a stage is not represented by a "dot2-like node but a vertical line like in the crossover. and for whatever band is selected, another section of the gui could show the pre-band effect parameters - like the compressot attack/release/ratio/etc. in this case.

...but implementing such a gui is a whole different story. first, the dsp has to be done. i think, we can have something to work with next week. i want to implement it as a baseclass MultiBandEffect (or something) and then we can have subclasses like MultiBandCompressor, MultiBandDistortion, etc...whatever comes up. the common MultiBand-splitting part will be factored out and can be re-used

**Comment by RobinSchmidt** [comment]
i'm currently setting up the general structure of the hierarchical bandsplitting with first order filters (which is the case which is known to work in providing perfect reconstruction) and later we may replace the 1st order filters with better ones (that's going to be the dsp-research part) without changing the big scheme of things

**Comment by RobinSchmidt** [comment]
ok, i have a bandsplitting algorithm that works by applying a pair of lowpass/highpass filters to a signal and then splits one of the two bands further. you can select whether the always the low or always the high band gets to split further. in case when you always split the low band, with 6 bands, a setup could look like:

![image](https://user-images.githubusercontent.com/1833099/33488644-9724484e-d6b1-11e7-8bad-0e3f16ef7c4f.png)

splitting always the low band further has the effect the lowest band hass the steepest accumulated filter slope (the low signal went through the greatest number of filters). you can also do it the other way around such that the high band goes to the greatest number of filters. 

the next step is to use a splitting strategy that distributes the accumulating filters slopes more evenly - i.e. each output should ideally go through the same number of filters. for example, with 8 bands, it should look more like a tree like this:

![image](https://user-images.githubusercontent.com/1833099/33488811-2a9a560e-d6b2-11e7-9e75-1cfc755fa1f6.png)

i have some algorithm based on recursion that would work like this when the number of bands is a power of two (as is the case here)....

**Comment by elanhickler** [comment]
...what? Can we have 3 or 7 bands? any arbitrary frequency split points?

**Comment by RobinSchmidt** [comment]
i want to convert the recursion into an iteration and somehow handle the cases when the number of bands is not a power of two. but there are may possibilities. for example, when i would only have 7 bands, i could leave out any of the 4 splitters in the last level (the binary tree would be incomplete). i could represent this by recombining either y0,y1 or y2,y3 ,...and the just not do the recombination literally but just leave the respective splitter out. ...not sure yet how to represent that algorithmically. ...actually i could also "recombine" y1-with-y2 - so for a 7-band splits there are either 4 or 8 possibilities from which i could choose.

of course, if you only need a fixed number of bands, these things could be hardcoded manually.

**Comment by RobinSchmidt** [comment]
> Can we have 3 or 7 bands? any arbitrary frequency split points?

with the "linear" structure above, yes - easily. with the tree structure...sort of, but we may have to recombine bands. it's not as straightforward. but that's what i want to work out next

**Comment by elanhickler** [comment]
I don't get why we are discussing some tree structure, you're just confusing me. So use the linear structure, we need simple arbitrary number of bands and frequency split points.

**Comment by RobinSchmidt** [comment]
run the "TestsRosicAndRapt" project. you will get output wav files where white noise has been split into 8 bands using the linear structure. the function bandSplittingMultiWay is run - you can set some options there - should be pretty self-explanatory. you could changes the setSlopeAccumulationMode function to see the difference between always splitting the low-band further and always splitting the high band further. it's easy to set up the computation structure for such a linear arrangement of splitters but i think in terms of the frequency responses, the tree would be more desireable

**Comment by elanhickler** [comment]
yes, but how do we know it's worth using the tree structure? the perceived quality difference may not be noticeable for the use case.

**Comment by RobinSchmidt** [comment]
what actually is the use-case in this case? is this for mastering purposes? it may or may not make a big difference. it probably becomes more relevant when the number of bands increases. in the case above, the high-band would have a final roll-off of 6 dB/oct but the low band 30 (6 above 100, 12 above 300, 18 above 1000, 24 above 3000 and 30 above 10000 - well, of course, that's only schematically thinking but you get the point)

**Comment by elanhickler** [comment]
this is using finite response filters, right?

**Comment by RobinSchmidt** [comment]
nope IIR. just a pair of 1st order lp/hp filters. i want to design a higher-order perfect reconstruction pair later (at least 2nd order) - but this will not change the overall structure. only the 2-way splitter code needs to be replaced

**Comment by elanhickler** [comment]
But we agreed to wanting to use FIR filters... what happened?

**Comment by RobinSchmidt** [comment]
oh - did we? your last post was actually

> I don't have an opinion on that

in response to my:

> i guess, it will be one of the two options:
> (1) FFT, spectral-splitting, several iFFTs
> (2) bandpass-bank with compensation (flattening) filter

...and then i actually wanted to go for (2) using engineers filter. ...but then i thought, perfect-reconstruction is too valuable a feature to give up so i wanted to try the 2-way hierarchical perfect-reconstruction tree first (the linear structure can actually be seen as a degenerate case of an incomplete tree, too)

**Comment by elanhickler** [comment]
Ok, whatever the case is, just give me something so we can test and see if it works. If you want to have a fixed number of bands but arbitrary split points, that could work. But fixed split points will not work for the client.

**Comment by RobinSchmidt** [comment]
as, said - you can test with the "TestsRosicAndRapt", how the splitters work suing white noise as input. i already have verified that the bands recombine perfectly to the original input (up to roundoff error). however, the separation may not be steep enough at the  moment - although maybe for a compressor we may not even want to steep separation anyway. some amount of band overlap may actually be desirable. however, it's probably too much at the moment

**Comment by elanhickler** [comment]
testing with white noise is a waste of time, I need to hear the end result. Gotta get this to my client. Once this is completed... or soon after, I can give you the money for you time spent on this AND the money for using your framework.

**Comment by RobinSchmidt** [comment]
> If you want to have a fixed number of bands but arbitrary split points, that could work. But fixed split points will not work for the client.

as said, with the linear structure, we have an arbitrary number of bands and arbitrary split points. with the tree, we now would have to choose a power of 2 for the number of bands, the splitting points ares still arbitrary.

> testing with white noise is a waste of time, I need to hear the end result. 

it's a very reasonable choice for a test-signal for the splitter itself during development because it contains all frequencies. what kind of input do you recommend? and you mean end-result as in the final multiband-compressor? do you want me to write the whole multiband compressor for you? i may do this but thought my part in this is only the splitter?

**Comment by elanhickler** [comment]
I'm not experienced enough to create a GUI. I already have a plugin that contains the EQ and limiter, now I need to add the multiband compressor, so you kinda have to do some of the work, enough so that I can add it into the plugin like I added the EQ and limiter.

**Comment by RobinSchmidt** [comment]
gui too? but the gui will not be shown to the client, right? that's what you said before. so it could be really basic?

**Comment by elanhickler** [comment]
yes

**Comment by RobinSchmidt** [comment]
hmm..ok...well, then i think, it is best when i make a multiband-compressor module for my chainer

**Comment by RobinSchmidt** [comment]
maybe i'll use the power-of-two tree algorithm internally and when you use, say 5, 6 or 7 bands, i'll internally recombine an appropriate number of bands. not optimal in terms of processing costs, but more straightforward to implement than an arbitrary number of bands tree. ...and may later be optimized

**Comment by RobinSchmidt** [comment]
...ooookayyy - there is now the MultiComp module available in Chainer (in the "Effects" section). it starts out with a single band. the NumBands slider chooses the number of bands from 1 to 16 (the code allows to change that upper limit by just changing a number at a particular place, but for convenience of the gui-work, i had to have an limit - may be lifted later by more complex gui work). the box below chooses which band's compression parameters (threshold, ratio, etc.) you currently see - and also the *upper* cutoff frequency of the band (i.e. the cutoff of the lowpass) - this implies that frequency setting the topmost band is actually irrelevant (if you have N bands, you have N-1 split frequencies).

i only did some very quick tests with a sinusoidal input. ...now more testing is needed. and probably some bug-fixing and refinement...

**Comment by elanhickler** [comment]
Soo, I don't understand why the multiband compressor has a selection between lowpass and highpass. Wouldn't that only be applicable to the first and last band? Shouldn't middle bands be using lowpass? Or how does this work?

**Comment by RobinSchmidt** [comment]
the selection between ""steep lowpass" and "steep highpass" mode selects, whether always the low or always the high band gets split further. see my scribble of the "linear structure" above. these are the two possible versions of it. say you have 3 bands - one split at 500 and one split at 5000 Hz. you could first split at 500 into a low (< 500) and high (> 500) band and then split the high band further at 5k. or you could first split at 5k and then split the low band further. the result is different final filter-slopes in the lowpass and highpass band. in the 1st case your lowpass would have a slope of 6 dB/oct at 500 hz and at 5k it would increase to 12 dB/oct. in the other case, your highpass would have a slope of 6 dB/oct at 5k and then steepen at 500 to 12 dB/oct

**Comment by RobinSchmidt** [comment]
well, no, there are no *literal* bandpasses. the bandpasses come about via combinations of lowpasses and highpasses

**Comment by RobinSchmidt** [comment]
i don't use literal bandpasses but instead a structure that splits a band into 2 - a highpass and a lowpass part. and then, any or both of these 2 bands can be split further - by the same kind of filter pair. this way makes it easy to ensure perfect reconstruction - it results automatically if your basic lowpass/highpass pair provides perfect reconstruction (which it always does in the case of a 1st order pair - but i may research into higher order pairs)

**Comment by RobinSchmidt** [comment]
my idea with the binary tree (as opposed to the linear structure) was to distribute the filter-slope accumulations more evenly between low and high bands. ...it sort of works but has its own problems. i've not yet made it available in the MultiComp, because i still have to handle non-power-of-two number of bands - but it should be easy in the simplest implementation

**Comment by RobinSchmidt** [comment]
btw. does your client want this for mastering applications?....because i may have an idea that might be much more efficient to compute and perhaps even better in the final result: replace the compressor with a leveller. in my current compressor code, i do quite expensive mathematical computations - convert from amplitude to dB, apply the compression function, convert back - which is no big deal in a single-band compressor, but when you use an array of them, it sums up...

**Comment by RobinSchmidt** [comment]
i came up with a quite simple and effective leveller formula. let *x* be the signal and *env* be the extracted envelope (by and envelope follower). modify the signal x according to the formula:
`x = a * x / (c + env)`
where *a* would be a target amplitude and *c* would be determined by some maximum gain that can be applied to a signal. for example, if c=0.1 you would never amplify any signal more than a factor of 10. you would replace compressor parameters "threshold" and "ratio" by leveller parameters "target-level" and "max-gain". ...would be easier to compute and perhaps nicer in the results

**Comment by elanhickler** [comment]
it is not for mastering, it is to recreate the sound of certain mobile device speakers.

btw, why doesn't the multiband compressor give access to a per-sample function? not usable without it.

**Comment by RobinSchmidt** [comment]
oh - you mean  the jura::MultiComp class? that it only has a block processing function? that's trivial to add, ..it's not yet totally complete

**Comment by elanhickler** [comment]
I know it's trivial to add. :)

I thought I was supposed to use the `MultiCompAudioModule` for M-Ref.

Well I'm using it.

**Comment by RobinSchmidt** [comment]
soooo - does it do what you want? does it work well for the intended purpose?

**Comment by elanhickler** [comment]
couldn't tell you! Waiting to hear back from the client.

**Comment by RobinSchmidt** [comment]
i guess, if it's for mobile speaker simulation, the issues i focused so much on (perfect reconstruction) don't really matter that much in this application....we'll see...in any case, we have a good basis for future mastering processors

**Comment by elanhickler** [comment]
the multicomp i think is pretty confusing.

If I want a 4 band split

200 1000 5000 10000 hz

what do splits do I set for bands and what should be set to lowpass and what should be set to highpass?

What if I want 3 bands? 400 4000 12000?

**Comment by RobinSchmidt** [comment]
> What if I want 3 bands? 400 4000 12000?

set NumBands to 3, select band 1 (via dropdown) and set it to 400, select band 2 and set it to 4000, select band and set it to 12000. only the sliders in the right-section are per-band parameters. the "Steep Lowpass", "Steep Highpass" is a global setting. it determines whether always the upper or always the lower band gets split further. so you don't select highpass or lowpass per band

**Comment by elanhickler** [comment]
so if lowpass is selected then

band 1 is from 0 to 400
band 2 is from 400 to 4000
band 3 is from 4000 to 12000

and 12000+ will just be "thru audio"?

**Comment by RobinSchmidt** [comment]
oh - now i'm confused myself. ...the setting above has actually 4 bands. when you have N split-frequencies, you have N+1 bands. consider the simple case of 2 bands - in this case you have 1 split-freq. as i said before:

> the NumBands slider chooses the number of bands [...] the box below chooses which band's compression parameters (threshold, ratio, etc.) you currently see and edit - and also the upper cutoff frequency of the band (i.e. the cutoff of the lowpass) - this implies that the frequency setting the topmost band is actually irrelevant

so, when you set NumBands to 3, you actually have only 2 split-frequencies. only SplitFrequency1 and 2 are relevant in this case. the 1st band is below f1, the 2nd band between f1 and f2, the 3rd band above f2. yes, it's quite confusing without a proper gui

**Comment by RobinSchmidt** [comment]
maybe i could gray out irrelevant sliders..

**Comment by elanhickler** [comment]
still don't understand lowpass vs highpass... does it have an audible effect? seems like it doesn't.

**Comment by RobinSchmidt** [comment]
ok, let's stick to the example: "steep lowpass" in this case means that the signal gets first split at 4000 into high and low and then low gets split further at 400 into mid and low. that means, the final lowpass lignal (below 400 hz) went through 2 lowpass filters (one at 4000 and one at 400). so it's steepness is fisrt 6 dB/oct (above 400) but at 4k, the slope steepens to 12 dB/oct. if you have more bands, the lowpass steepens more and more towards high frequencies, whereas the highpass is 6 dB/oct all the way down to zero. when you select "steep highpass", it's the other way around. if it makes no audible difference, it's actually a good thing - because both versions are complementary compromises. if they are roughly equal, it means the compromises are actually not that bad

**Comment by WelchCompositions** [comment]
Hey guys! I think I want to discuss a budget for making a nice visual GUI for the multiband compressor similar to the EQ. Robin, can you and Elan let me know what that would cost to implement into our current plugin suite? Thanks!

**Comment by elanhickler** [comment]
robin, don't forget to respond to this.

I'll do the implementation, it will be covered by your budget. Robin will have to give you an estimate on coding the GUI.

**Comment by RobinSchmidt** [comment]
as said in the email, my usual hourly rates are either 30€ for stuff that my library benefits from or 50€ for other stuff. this falls into the first category. i would estimate maybe 2 or 3 workdays for such a thing to implement., so maybe between 500...700€

**Comment by RobinSchmidt** [comment]
you send me this as example:
![image](https://user-images.githubusercontent.com/1833099/34782492-2bf41d8c-f629-11e7-8fcf-44cc891e5655.png)
i suppose the vertical lines work similar to those in my crossover and the nodes/dots adjust the compressor threshold for each band? ...i wonder the, if ratio could also somehow be represented in the plot - maybe with a vertical line representing the input dB-over-threshold value that corresponds to 1 dB output-dB-over-threshold. ...but we may need to think some more about this - maybe someone has a better idea

**Comment by WelchCompositions** [comment]
Hey Robin, Let's go ahead and pull the trigger based on your estimate. Ive found 90% of multiband compressors tend to have similar layouts/GUI so there should be lots of inspiration for ya if you do a google image search. I just grabbed the most generic one I could find.

**Comment by WelchCompositions** [comment]
Im open to alternative designs but lets not reinvent the wheel unless there is a good reason to. Once you find a layout you like and want to move forward with just post an image of that here for approval. Thanks!

**Comment by WelchCompositions** [comment]
@elanhickler Im thinking maybe we do a tab design to go between the different effects in the suite? Im open to alternative designs but that seems cleanest

**Comment by elanhickler** [comment]
yeah i can do tabs. I'd think you'd want to interact with multiple fx easily at once, but maybe not. So tabs will work.

**Comment by WelchCompositions** [comment]
From my experience, for me at least, matching the sound profile tends to start with EQ which gets it pretty darn close, then compression to match the ability for certain frequencies to pop out, then using the limiter and/or gain to bring the signal higher or lower if needed.

I'm open to suggestions for other layouts if you come up with something graceful but tabs seem to match my expected workflow

**Comment by RobinSchmidt** [comment]
> matching the sound profile tends to start with EQ which gets it pretty darn close, then compression to match the ability for certain frequencies to pop out

have you considered replacing multiband compression with dynamic eq? this would bypass any issues associated with band recombination (although it may introduce others)...but when there's an eq in the chain anyway, it seems like a more straightforward addition than a multiband compressor with similar effect.

i don't have any ready-made dynamic-eq code around, though - just something to think about

**Comment by WelchCompositions** [comment]
Dynamic EQ _could_ potentially solve the problem but I'm not experience with it enough to know for certain. When I was working with just EQ my ear told me I needed a compressor in the chain to solve the problems I was hearings. Its possible dynamic EQ could also solve it. Though since we are already in route to a multiband compressor and I feel more comfortable/confident in it I think we should go that route :)

**Comment by WelchCompositions** [comment]
@RobinSchmidt Hey Robin! Just wanted to check in to see if you've gotten rolling. Looking forward to seeing what you create!

**Comment by RobinSchmidt** [comment]
i did not yet get around this because i'm currently reworking our parameter handling infrastructure (i've broken some things with the addition of a recent new feature). i think, i will need a few days. @elanhickler - i hope that tomorrow i can make it all work again, then i will implement the curve options. when this is done, i can look into something new

**Comment by WelchCompositions** [comment]
ok, no rush. There are some other delays on this project so we have plenty of breathing room

**Comment by elanhickler** [comment]
for the interface I propose two tabs, one for EQ, one for Compressor, but always have visible the limiter whatever other basic controls that don't take up much space.

**Comment by RobinSchmidt** [comment]
> ok, no rush

ok, good to hear. still working on some details of the parameters

**Comment by elanhickler** [comment]
jim I think Welch needs his GUI, stop working on your dream synth! 😄

**Comment by RobinSchmidt** [comment]
ahhh...but my dream synth!!! <3 ...but yeah...i think, i need to think about some design issues anyway before continuing with this (especially with regard to a polyphonic mod-system)

so, ok yes, i'll start working on this multicomp gui this week...maybe i can do this in parallel.

i'll also refactor my messy coordinate system drawing and function plotting code - this will be helpful then for implementing a multicomp gui

**Comment by RobinSchmidt** [comment]
yes, i'm currently refactoring (almost rewriting) my horrible drawing and plotting code. the old code is so clunky and inflexible - it needs to be redone before writing any new code on top of that

**Comment by RobinSchmidt** [comment]
ok, the refactoring of general plotting facilities is more or less done. i'm now putting the new code to use - first for the ladder, then for the eq

**Comment by elanhickler** [comment]
What's better about the new plotting? Will my synths benefit? (since I use the EQ)

**Comment by RobinSchmidt** [comment]
look at class rsLadderPlotEditor (the new one) vs class LadderSpectrumEditor (the old one, to be deleted as soon as transition is complete). i'm achieving more or less the same thing with much less code - because a lot of the drawing code has been factored out in a reusable way. everything plotting related is much more flexible and easy to reuse now.

the eq gui still has to be updated. i hope to be able to significantly reduce the code size there, too - which will make it easier to create new guis with similar functionality

**Comment by RobinSchmidt** [comment]
ahh...and maybe we can soon have filled graphs, too. they kinda look better (but are also more expensive)

**Comment by RobinSchmidt** [comment]
yay!
![image](https://user-images.githubusercontent.com/1833099/36217140-f67d07ba-11b0-11e8-9334-7fa231a752d3.png)

**Comment by elanhickler** [comment]
how do you gert that report?

**Comment by RobinSchmidt** [comment]
i use tortoise git. it always gives me this on commit

**Comment by WelchCompositions** [comment]
While the details are a bit beyond me, nice work! Always exciting to hear progress.

**Comment by RobinSchmidt** [comment]
i started working on the multicomp gui
![image](https://user-images.githubusercontent.com/1833099/36500240-0c868b38-1744-11e8-9d49-cc35f6924857.png)

**Comment by elanhickler** [comment]
amazing!

**Comment by WelchCompositions** [comment]
looking good!

**Comment by RobinSchmidt** [comment]
hey jim - sorry for not making progress this week. i just discovered a new  (and, as far as i know, largely unexplored) algorithm for sound synthesis and was totally carried away exploring this (writing lots of novel code). if you are curious, what it is about, check out this thread:

https://github.com/RobinSchmidt/RS-MET/issues/194

this was a very productive and creative week in this regard. i just had to do this as long as i'm in the flow. do you have any time-schedule/deadlines for when you you need the multicomp gui?

**Comment by WelchCompositions** [comment]
I don't but all my financial investment in this (it's been a must year R&D process) is tied up until it's finished. So no rush but I want to make sure it does get done and doesn't sit around untouched for too long. Do you think we can have it wrapped up this month?

**Comment by RobinSchmidt** [comment]
yes, ok - i think, that should be doable

**Comment by RobinSchmidt** [comment]
did you already create a lot of presets? because i'm currently changing a couple of things that may break presets. the way, i handle the number of bands is a bit awkward and writing a proper gui on top makes it even more awkward, so it's a good idea to change that now. if all goes well, my changes will also lift the limit of 16 bands at maximum to a truly arbitrary number. i could write some preset conversion code to make it backward compatible, if necessarry - that's why i ask

**Comment by WelchCompositions** [comment]
nope all presets made we just testing so feel free to break it!

**Comment by WelchCompositions** [comment]
Thanks Robin! @elanhickler will you be ready to roll once Robin wraps his phase?

**Comment by elanhickler** [comment]
Yes

**Comment by RobinSchmidt** [comment]
i'm just changing the whole multiband-effect infrastructure in a way that will me allow to select the kind of multiband-effect dynamically at runtime. that means, it will not just be a multiband-compressor but a multiband-anything - for any kind of effect for which i already have a single-band implementation - which are a lot (i will then just need to add one or two lines of code to make a new type of effect available with the multiband effect framework). the user can then select the kind of effect via dropdown menu. it's a bit more complex to lay this infrastructure down but it will pay off may be even less code to write, because i may directly re-use all my single-band effects (and their guis). i think, i should have thought of that right from the beginning anyway

**Comment by RobinSchmidt** [comment]
yay! the basic system is in place! now debugging and polishing...

![image](https://user-images.githubusercontent.com/1833099/38706839-0157eb98-3eaf-11e8-84ab-a18c81d7fb86.png)

**Comment by RobinSchmidt** [comment]
ooookay - i think, it is now in a state where it could make sense for elan to wrap it and to put it actually to use. there still some features missing (like draggability of the vertical lines in the plot and frequency response curves) but the general system is done. the hardest part was actually not so much related to gui programming per se but to create a framework where the user can freely add and remove bands at arbitrary positions.

at the moment, the split frequencies must be adjusted via the SplitFrequencyX sliders. the box that says "nothing selected" can be ignored for the moment. this is where later the user can select which kind of multiband-effect it should be (currently, there's only the compressor - but it will be easy to add more).

at the moment, there's nothing in the code that enforces the SplitFrequencies to be in ascending order, but i'm considering to enforce this later (for various reasons related to extensions i have in mind), so when you create presets, please make sure to use ascending split-frequencies. ...oh, and it needs testing...

**Comment by RobinSchmidt** [comment]
i just did some testing and fixed a little bug. also, for technical reasons, it's currently not possible to remove the rightmost band. these are things that need to be solved for an actually viable product to be released to the public....later....

**Comment by WelchCompositions** [comment]
Awesome work! If Elan can get it wrapped into something for me to test and verify its functioning (or if you're confident its solid) I can shoot you over payment. @elanhickler Let me know if you need anything and a rough timeline for when you can get this in place. Thanks!

**Comment by RobinSchmidt** [comment]
it's not yet completely finished. i mean, it seems to work in the sense that we as devs can reasonably work with it. but i actually like to polish it some more so as to make it actually viable for an end-user, too. but this polishing and addition of features can be done in parallel or even after elan does the wrapping for you. i've taken some days off from it the last few days for learning to use the sage math software system and adding more (math-related) code to my library. but i'll pick it up soon again to fill the gaps...

**Comment by WelchCompositions** [comment]
@elanhickler Are you currently at work on this?

**Comment by elanhickler** [comment]
Sorry for taking so long on this. Yes, Saturday and Sunday I'll dedicate to this project. I hope I can get it done then too because I need to move on. I don't think I'll be looking for programming work after this actually.

**Comment by RobinSchmidt** [comment]
>  Actually Robin I hope you can take over this project after I complete it initially.

does this mean, i will have to build mac versions and installers? i'm not very good at that. but i'd like to improve the dsp side in the future - i have some ideas how to improve the band-splitter

**Comment by elanhickler** [comment]
no im talking more like... if Mr. Welch needs something some months from now.

**Comment by RobinSchmidt** [comment]
implementing new features, especially on the dsp side, is always welcome. i'm just not very keen on maintaining the build system, deployment process, etc. - i'm the algorithms-guy :-)

**Comment by elanhickler** [comment]
robin, could you quickly remind me how to use the mutliband comp? what classes do I use?

**Comment by RobinSchmidt** [comment]
oh - yes, it's now jura::MultiBandEffect and jura::MultiBandEffectEditor

**Comment by elanhickler** [comment]
I think I got it working. @WelchCompositions i just need to spend some time on organizing the GUI and you should have a working prototype soon.

**Comment by elanhickler** [comment]
Robin, what did you change that would make an extra Gain And Mix panel appear? What should I do?

![image](https://user-images.githubusercontent.com/9116575/40029088-732bb1b8-5797-11e8-91ac-8045412ddeda.png)

**Comment by RobinSchmidt** [comment]
are you using jura::LimiterAudioModule as is? or do you use a subclass? or your own limiter class based on my dsp? i have that limiter also available in ToolChain and there, its gui looks like this:
![image](https://user-images.githubusercontent.com/1833099/40029924-9290e8ac-57e7-11e8-9fb0-ea3639ec46fd.png)

**Comment by elanhickler** [comment]
I'm using a subclass to override resized.

Ok I got rid of the Gain And Mix by doing `othersLabel->setVisible(false);`

I'm not using the original resized, it's overridden, not sure why it's being called and creating panels and text.

**Comment by elanhickler** [comment]
`guiLayoutRectangles.clear();` works

**Comment by elanhickler** [comment]
http://www.elanhickler.com/transfer/Soundemote/M-Ref%20Dev%20Installer.exe
http://www.elanhickler.com/transfer/Soundemote/M-Ref%20Dev.pkg

**Comment by WelchCompositions** [comment]
no worries! I may not be able to check this until early next week as Im out of town but excited to see it!

**Comment by WelchCompositions** [comment]
Hey @elanhickler  Got some time to put it through the ringer. Really nice work to you and Robin! A couple small bugs/issues I found:

Dry/Wet doesn't function as expected. 100% dry should sound the same as bypassed but it does not.

Switching stereo mode in the EQ crashed Logic Pro

@RobinSchmidt Im not seeing expected visual feedback on multi band compressor. Adding and removing bands is amazing and seems to be working great. However, Im not seeing feedback from the settings of each band or how the audio is coming into each band. While its functional technically the lack of visual feedback will make it challenging to use in practice. Below is an example of the feedback Im referring to. What do you think?

http://www.modernmetalproduction.com/wp-content/uploads/2017/08/m4newgui_screenshot6.png

**Comment by elanhickler** [comment]
I think he means he wants a real time meter repaint method for each band to show RMS or amplitude

**Comment by RobinSchmidt** [comment]
>  Im not seeing expected visual feedback on multi band compressor. Adding and removing bands is amazing and seems to be working great. However, Im not seeing feedback from the settings of each band or how the audio is coming into each band. While its functional technically the lack of visual feedback will make it challenging to use in practice. Below is an example of the feedback Im referring to. What do you think?

ahh - i just noticed this post. so you mean a visual feedback inside the plot? yes, this would be nice to have. it's not supposed to be finished anyway since the frequency response curves are still missing. it's currently just to select/add/remove bands. but yes, in order to make it a viable gui, things like that should be added

**Comment by RobinSchmidt** [comment]
what kind of product should this actually be? elan told me that the whole gui is actually just needed at the developers side, not in the actual product - which should just let the user select between a bunch of fixed preset settings (to be created via the dev-version which has some sort of (workable but not necessarily polished) gui  to create the presets). is that right?

i have just gotten around to hack my work hours i have spent on this into an excel sheet (from pencil/paper notes) and i'm now at 592.50€ in total (for dsp and gui code). if you want me to and your budget allows, i could work more on the gui to include the visual feedback

**Comment by WelchCompositions** [comment]
That is correct it's just for the developer to create presets. My concern is creating those presets will be challenging with limited feedback. While it doesnt need to be beautiful it would be good to be easy to use. Any estimate on the rough final cost when you add the other elements? I'd like to go for it but just want to get an idea of what the total investment will be. This is the product it is for, it's the software I use to emulate other devices http://www.m-ref.com/

**Comment by WelchCompositions** [comment]
That sounds good Robin, let’s do it. Elan, do you know if you and I still have a balance? Just trying to get a balance for total cost

**Comment by elanhickler** [comment]
$300 on completion

**Comment by WelchCompositions** [comment]
Thanks Elan let's rock and roll then! :)

**Comment by WelchCompositions** [comment]
Hey Robin, just wanted to check in on progress :)

**Comment by RobinSchmidt** [comment]
i'm just finishing off some work for elan - will probably be done tomorrow - then i'll work on this

**Comment by WelchCompositions** [comment]
ok sounds good!

**Comment by RobinSchmidt** [comment]
ok - there are some gain indicators now. ..but it's still horribly inefficient - the whole background is redrawn for every frame (when actually only the gain-bars need to be redrawn). i want to optimize that. and add the freq-response curves next...

**Comment by WelchCompositions** [comment]
awesome! Thanks for the update :)

**Comment by WelchCompositions** [comment]
Hey Robin! Just a friendly check in, Im still doing back and forths on the enclosure that should likely last another month or two so no rush but want to make sure this stays active and doesn't get lost in the weeds :)

**Comment by RobinSchmidt** [comment]
ok, good to hear. i actually have the optimization i was talking about finished - now only the freq-response curves are missing. maybe i'll look into that next week. i'm currently working on my modular system again, which i left off few years ago.

after the freq-response curves are done, i may want to look into improving the frequency splitter code to use higher order filters. that's then a dsp research task again

**Comment by RobinSchmidt** [comment]
we have freq-response curve plots now. now we can finally see the difference between "steep highpass" and "steep lowpass" mode.

**Comment by WelchCompositions** [comment]
awesome! So does that complete the visual feedback request or are there any pieces missing?

**Comment by RobinSchmidt** [comment]
i think the visual feedback is completed now - but that's actually your call ;-) ...but i think, there's still room for cosmetic improvements (the colors of the plots don't look very attractive at the moment, in my opinion). and i want to make these vertical lines available for grab-and-drag like in my old crossover plugin

**Comment by WelchCompositions** [comment]
Those sound like acceptable shortcomings for my usage. Shoot me an invoice to jim@m-ref.com and send Elan anything he may need if you could @elanhickler  are you ready to pick this up and take it home?

**Comment by WelchCompositions** [comment]
@elanhickler checking back in to see if we can finish this project up, thanks!

**Comment by RobinSchmidt** [comment]
>  Shoot me an invoice to jim@m-ref.com and send Elan anything he may need

it's all right here in the codebase. with regard to the invoice -  i'll get back to you about that soon. ...have to transfer my work-hours from paper to pc. by the way - i have already done some possibly relevant dsp research on my own (without charging anyone):

https://github.com/RobinSchmidt/RS-MET/issues/219

but it's all still in preliminary research state and not yet incorporated

**Comment by elanhickler** [comment]
ill see if i can get this done tomorrow. i leave for california so if I don't get it done it will be another 5 days.

can you remind me what I should be doing tomorrow?

**Comment by WelchCompositions** [comment]
Sorry Elan I just saw this. Incorporate Robins new multi band compressor and then I can run through all the functionality and see if there are any remaining issues/bugs and if not we can then make sure its all buttoned up for distribution!

**Comment by RobinSchmidt** [comment]
yes - this output volume indication is the new visual feedback feature, we were talking about. what do you mean by "being concerned about"? i thought, you'd just grab the update and re-build the plugin(s)?

showing the spectrum...yeaaa...certainly possible - but it would require quite a bit work and it's not currently high on my own priority list (but if someone wants to commission this at some point, why not....)

**Comment by elanhickler** [comment]
ok i got tis working. I'm good for now.

**Comment by elanhickler** [comment]
many build errors on Mac! Can't build M-ref. Errors don't look too difficult.

**Comment by elanhickler** [comment]
Here's the windows builds for now:
http://www.elanhickler.com/transfer/Soundemote/m-ref%20builds.zip

**Comment by RobinSchmidt** [comment]
ok - mac build errors fixed.

**Comment by elanhickler** [comment]
I'm getting RAPT::rsRoundToInt<double>(double) used in function jura::rsNodeBasedFunction::setupNodeParameterCallbacks

symbol not found for architecture i386

**Comment by RobinSchmidt** [comment]
on mac? i just tried the 32-bit build of toolchain on mac as well and it worked.

**Comment by RobinSchmidt** [comment]
> restart your computer, sounds like the file is in use / locked erroneously.

ok - that seems to have worked. i can build win32 now. so - where do you get this error?

**Comment by elanhickler** [comment]
now it looks like x86_64 is giving me linker error.

same error RAPT::rsRoundToInt(double) used in function jura::rsNodeBasedFunction::setupNodeParameterCallbacks

**Comment by elanhickler** [comment]
RAPT errors usually mean you didn't create an instance of the specific template function (double in this case). I deleted the use of that function in setupNodeParameterCallbacks and it built successfully. I'm not using node based functions for M-ref... I think......? So M-ref should work still I hope.

**Comment by RobinSchmidt** [comment]
> RAPT errors usually mean you didn't create an instance of the specific template function (double in this case)

indeed. but why does it work here, then? maybe sometimes the compiler creates a template instantiation itself and sometimes it doesn't? but that would be weird

**Comment by elanhickler** [comment]
no longer relevant / old discussion
